### PR TITLE
Complex is now serialized as two value array rather than str

### PR DESCRIPTION
**Pull request by loichuder**

### CODE REVIEW

**Comment by telamonian** [comment]
This PR seems good like a good idea overall. I'm a touch worried though about it will interact with the existing shape inference and shape coercion code. Basically, I'm concerned that something will end up treating eg a `(3, 3, 3)` array of complex as a `(3, 3, 3, 2)` array of reals.

I'll trace through the code until I'm satisfied this won't cause an issue (or until I find an issue). Alternatively, I suppose we could firm up the unittests for shape inference and coercion instead. That would probably also be sufficient to ease my worried mind

**Comment by loichuder** [comment]
I checked and everything works as intended for complex scalars, 1D arrays and nD arrays.

I think it stems from the fact that the shape inference and shape coercion code uses the NumPy arrays that have the appropriate dimensions and have no knowledge about how the value is then serialized.

**Review by telamonian — CHANGES_REQUESTED** [review]
> I think it stems from the fact that the shape inference and shape coercion code uses the NumPy arrays that have the appropriate dimensions and have no knowledge about how the value is then serialized.

Sounds about right. If you add whatever update is needed to properly display these 2-tuple complexes on the frontend, we can call this PR done.

I would also like to see at least one unittest that covers the `min_ndim` query parameter, but I wouldn't say that's a blocker for this PR

File: .eslintrc.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ module.exports = {
     // 'jest',
   ],
   rules: {
-    '@typescript-eslint/no-floating-promises': ['error', { ignoreVoid: true }],
+    '@typescript-eslint/no-floating-promises': ['warn', { ignoreVoid: true }],
```
Comment (loichuder) [inline .eslintrc.js:30]: The `pre-commit` hook was preventing me from commiting `dataset.ts` due to this. I set it to warning until we fix the errors.

File: .eslintrc.js [hunk 2]
```diff
@@ -65,6 +65,10 @@ module.exports = {
     'prefer-const': 'off',
     // 'react/prop-types': 'warn'
   },
+  env: {
+    browser: true,
+    node: true,
```
Comment (loichuder) [inline .eslintrc.js:70]: Without this, `document` is treated as undef and `module` as well.

File: src/hdf.ts [hunk 3]
```diff
@@ -100,12 +101,12 @@ export function hdfContentsRequest(
 export function hdfDataRequest(
   parameters: IDataParameters,
   settings: ServerConnection.ISettings
-): Promise<number[][]> {
+): Promise<number[][] | Complex[][]> {
```
Comment (loichuder) [inline src/hdf.ts:104]: The main change is here

File: src/hdf.ts [hunk 4]
```diff
@@ -162,11 +163,11 @@ export async function hdfApiRequest(
     if (data.length > 0) {
       try {
         // HTTPError on the python side adds some leading cruft, strip it
-        json = JSON.parse(data.substring(data.indexOf("{")));
-      } catch (error) {}
+        json = JSON.parse(data.substring(data.indexOf('{')));
+      } catch (error) {} // eslint-disable-line no-empty
```
Comment (loichuder) [inline src/hdf.ts:167]: Again, I had to do this to make the pre-commit happy

File: src/dataset.ts [hunk 5]
```diff
@@ -180,28 +181,35 @@ export abstract class HdfDatasetModel extends DataModel {
...
+    params: IDataParameters
+  ): Promise<number[][] | string[][]> {
     try {
-      return await hdfDataRequest(params, this._serverSettings);
+      const data = await hdfDataRequest(params, this._serverSettings);
+      const { dtype } = this.meta;
+      if (isComplexArray(data, dtype)) {
+        return data.map(inner => inner.map(convertToStr));
+      }
+      return data;
```
Comment (loichuder) [inline src/dataset.ts:212]: Main change is here

**Review by loichuder — COMMENTED** [review]
Formatting lead to heavy diff. I tried to underline the main changes when needed.

**Review by telamonian — APPROVED** [review]
LGTM!

**Pull request merged**